pkg/preparer/result_set_test.go

```diff
@@ -24,36 +26,68 @@ func TestZipBasic(t *testing.T) {
 	}
 
 	pairs := ZipResultSets(intent, reality)
-	ValidatePairList(t, pairs)
-	for _, pair := range pairs {
-		switch pair.ID {
-		case "bar":
-			Assert(t).IsNotNil(pair.Intent, "bar pair should have intent")
-			Assert(t).IsNotNil(pair.Reality, "bar pair should have reality")
-		case "baz":
-			// the nil pointer inside ManifestPair is getting boxed into an
-			// interface when passed as an argument
-			// therefore, Assert(t).IsNil doesn't work here
-			if pair.Intent != nil {
-				t.Error("baz pair should not have intent")
-			}
-			Assert(t).IsNotNil(pair.Reality, "baz pair should have reality")
-		case "foo":
-			Assert(t).IsNotNil(pair.Intent, "foo pair should have intent")
-			if pair.Reality != nil {
-				t.Error("baz pair should not have reality")
-			}
-		}
-	}
+	ValidatePairList(t, pairs, intent, reality)
 }
 
-func ValidatePairList(t *testing.T, l []ManifestPair) {
+func ValidatePairList(t *testing.T, l []ManifestPair, intent, reality []kp.ManifestResult) {
 	for _, pair := range l {
+		// fuzz check #1: all three fields of the Pair should have the same ID
 		if pair.Intent != nil {
 			Assert(t).AreEqual(pair.ID, pair.Intent.ID(), "intent manifest should have had same ID")
 		}
 		if pair.Reality != nil {
 			Assert(t).AreEqual(pair.ID, pair.Reality.ID(), "reality manifest should have had same ID")
 		}
+
+		// fuzz check #2: if Pair.Intent is non-nil, then it should be in the
+		// source intent list, and vice versa
+		var intentFound bool
+		for _, m := range intent {
+			if pair.ID == m.Manifest.ID() {
+				// cannot use Assert.IsNil against a pointer
+				if pair.Intent == nil {
+					t.Errorf("source intent list and pair contained matching id %q, but pair did not contain intent manifest", pair.ID)
+				}
+				intentFound = true
+			}
+		}
+		Assert(t).IsTrue(pair.Intent == nil || intentFound, "should have found at least one manifest in the source intent list")
+
+		// fuzz check #3: same as #2, but for reality
+		var realityFound bool
+		for _, m := range reality {
+			if pair.ID == m.Manifest.ID() {
+				// cannot use Assert.IsNil against a pointer
+				if pair.Reality == nil {
+					t.Errorf("source reality list and pair contained matching id %q, but pair did not contain reality manifest", pair.ID)
+				}
+				realityFound = true
+			}
+		}
+		Assert(t).IsTrue(pair.Reality == nil || realityFound, "should have found at least one manifest in the source reality list")
 	}
 }
+
+func TestZipFuzz(t *testing.T) {
+	for i := 0; i < 500; i++ {
+		intent := generateRandomManifestList(100)
+		reality := generateRandomManifestList(100)
```

Comment:
Minor nit: `generateRandomManifestList()` is very unlikely to generate two lists that have any overlap.